students/ArchWu/Session6/mailroom.py

```diff
@@ -0,0 +1,186 @@
+#!/usr/bin/env python3
+import sys
+import tempfile
+from textwrap import dedent
+
+import math
+from operator import itemgetter
+
+#donations[]
+#donors {donor1:donations1, donor2:donations2}
+donors = {}
+# The data struction used in this program, a list of tuples
+def get_donors():
+    """ Initializer of the dictionary containing informations of donors """
+    return {'william gates iii': ("William Gates III", [653772.32, 12.17]),
+            'jeff bezos': ("Jeff Bezos", [877.33]),
+            'paul allen': ("Paul Allen", [663.23, 43.87, 1.32]),
+            'mark zuckerberg': ("Mark Zuckerberg", [1663.23, 4300.87, 10432.0]),
+            }
+
+def find_donor(name):
+    """ Find the donor regardless of upper/lower cases and ignores extra spaces """
+    key = name.strip().lower()
+    return donors.get(key)
+
+def add_donor(name):
+    """ Adds a new donor """
+    name = name.strip()
+    donor = (name, [])
+    donors[name.lower()] = donor
+    return donor
+
+def list_donors():
+    """
+    Create a list of the donors as a string, so they can be printed
+
+    Not calling print from here makes it more flexible and easier to
+    test
+    """
+    listing = ["Donor list:"]
+    for donor in donors.values():
+        listing.append(donor[0])
+    return "\n".join(listing)
+
+def add_donation(name, donation):
+    """ Adds a donation to an existing donor """
+    donors[name.strip().lower()][1].append(donation)
+    return
+
+def generate_donor_report():
+    """ Make a report on donations received by a style as specified """
+    sorted_donors = sorted(donors.items(), reverse = True, key = lambda item: (sum(item[1][1])))
+    report_rows = [print("{0:25s}   {1:10.2f}   {2:9d}   {3:11.2f}".format(*gen_stats(donor))) for donor in sorted_donors]
+    report = []
+
+    report.append("{:25s} | {:11s} | {:9s} | {:12s}".format("Donor Name",
+                                                            "Total Given",
+                                                            "Num Gifts",
+                                                            "Average Gift"))
+    report.append("-" * 66)
+
+    for row in report_rows:
+        report.append(row)
+    return "\n".join(report)
+# def generate_donor_report():
```

Comment:
It would be good to remove out this commented code since it just clutters this file